pennylane/control_flow/while_loop.py

```diff
@@ -178,38 +306,94 @@ def _call_capture_disabled(self, *init_state):
 
         return fn_res
 
+    def _handle_error(self, e: ValueError):
+        """Handle any ValueError's raised by the creation of the jaxpr, adding information to any error
+        about 'Incompatible shapes for broadcasting'."""
+        import jax  # pylint: disable=import-outside-toplevel
+
+        if "Incompatible shapes for broadcasting" in str(e) and jax.config.jax_dynamic_shapes:
+            closures = (self.body_fn.__closure__ or ()) + (self.cond_fn.__closure__ or ())
+            if any(_has_dynamic_shape(i.cell_contents) for i in closures):
+                msg = (
+                    "Detected an attempt to combine arrays with different dynamic shapes. "
+                    "\nThis also may be due to a closure variable with a dynamic shape."
+                    " Try promoting the closure variable with the dynamic shape to being an explicit argument. "
+                )
+                if self.allow_array_resizing is True:
+                    msg += "\nThis may also be due to allow_array_resizing=True. Try with allow_array_resizing=False instead."
+            else:
+                msg = (
+                    "Detected an attempt to combine arrays with two different dynamic shapes. "
+                    "To keep dynamic shapes matching, please specify ``allow_array_resizing=False`` to ``qml.while_loop``."
+                )
+            raise ValueError(msg) from e
+        raise e
+
+    def _get_jaxprs(self, init_state, allow_array_resizing):
+        import jax  # pylint: disable=import-outside-toplevel
+
+        flat_args, in_tree = jax.tree_util.tree_flatten(init_state)
+        tmp_array_resizing = False if allow_array_resizing == "auto" else allow_array_resizing
+        abstracted_axes, abstract_shapes, shape_locations = loop_determine_abstracted_axes(
+            tuple(flat_args), allow_array_resizing=tmp_array_resizing
+        )
+
+        flat_body_fn = FlatFn(self.body_fn, in_tree=in_tree)
+        flat_cond_fn = FlatFn(self.cond_fn, in_tree=in_tree)
+
+        if abstracted_axes:
+            new_body_fn = _add_abstract_shapes(flat_body_fn, shape_locations)
+            dummy_init_state = [_get_dummy_arg(arg) for arg in flat_args]
+        else:
+            new_body_fn = flat_body_fn
+            dummy_init_state = flat_args
+
+        try:
+            jaxpr_body_fn = jax.make_jaxpr(new_body_fn, abstracted_axes=abstracted_axes)(
+                *dummy_init_state
+            )
+            jaxpr_cond_fn = jax.make_jaxpr(flat_cond_fn, abstracted_axes=abstracted_axes)(
+                *dummy_init_state
+            )
+        except ValueError as e:
+            self._handle_error(e)
+
+        validation = _validate_no_resizing_returns(jaxpr_body_fn.jaxpr, shape_locations)
+        if validation:
```

Comment:
`if validation` doesn't really sound like an error was raised to me - maybe the variable could be `error_raised = _validate_no_resizing_returns` so that it becomes `if error_raised`? 

Or else just a comment like in the `for_loop` PR, so that its easier to see what's happening.